### ISSUE TITLE
NCZarr - Netcdf Support for Zarr

### ISSUE THREAD

**Issue by DennisHeimbigner** [zarr-developers/zarr-specs#41]
I am moving the conversation about NCZarr to its own issue. See Issue https://github.com/zarr-developers/zarr/issues/317 for initial part of this discussion.

**Comment by alimanfoo** [comment]
Surfacing here [notes on the NetCDF NCZarr implementation](https://drive.google.com/file/d/1UUGcQMpWqKllMdRFCu97CoL7fB_GWXvg/view), thanks @DennisHeimbigner for sharing.

**Comment by alimanfoo** [comment]
Also relevant here, [documentation of xarray zarr encoding conventions](http://xarray.pydata.org/en/latest/internals.html#zarr-encoding-specification), thanks @rabernat.

**Comment by rsignell-usgs** [comment]
@DennisHeimbgner: It looks like Unidata's Netcdf C library can now [read data with the xarray zarr encoding conventions](https://github.com/Unidata/netcdf-c/pull/1952), right?

@rabernat, should I raise an issue for xarray to also support the [Unidata NcZarr](https://www.unidata.ucar.edu/blogs/developer/entry/overview-of-zarr-support-in) conventions?

**Comment by WardF** [comment]
The ability to read xarray is in the main branch, and will be in the upcoming 4.8.1 release. I am shaving the yak to get our automated regression and integration test infrastructure back up and running but we hope to have 4.8.1 out shortly.

**Comment by rabernat** [comment]
> @rabernat, should I raise an issue for xarray to also support the [Unidata NcZarr](https://www.unidata.ucar.edu/blogs/developer/entry/overview-of-zarr-support-in) conventions?

I see this as very difficult. The reason is that the ncZarr conventions use files outside of the zarr hierarchy. We would probably need to implement a compatibility layer as a third-party package, similar to h5netcdf.

**Comment by shoyer** [comment]
One thing I'll note on Xarray's convention for the Zarr is that we will likely change things in the near future to always write and expect "consolidated metadata" (see https://github.com/pydata/xarray/issues/5251). This is almost completely backwards compatible, but if NcZarr writes these consolidated metadata fields in Xarray compat mode we could load these Zarr stores much quicker in Xarray.

Consolidated metadata would probably be a nice feature for NcZarr, too, because it reduces the number of files that need to be queried for metadata down to only one. I think there was a similar intent behind the `.nczgroup` JSON field. Consolidated metadata is sort of a super-charged version of that.

**Comment by DennisHeimbigner** [comment]
NCZarr get similar improvement by doing lazy read of metadata objects. That is one problem with _ARRAY_DIMENSIONS -- it requires us to read all attributes even if otherwise unneeded. NCZarr avoids this by keeping the dimension names separate.
As for consolidated metadata, I assume you are NOT saying that any pure zarr container that does not contain the consolidated metadata will be unreadable by Xarray.

**Comment by shoyer** [comment]
> NCZarr get similar improvement by doing lazy read of metadata objects. That is one problem with _ARRAY_DIMENSIONS -- it requires us to read all attributes even if otherwise unneeded. NCZarr avoids this by keeping the dimension names separate.

In Xarray, we have to read nearly all the metadata eagerly to instantiate `xarray.Dataset` objects.

> As for consolidated metadata, I assume you are NOT saying that any pure zarr container that does not contain the consolidated metadata will be unreadable by Xarray.

This is correct, you don't need to write consolidated metadata. But if you do, Xarray will be able to read the data much faster.

As for whether netCDF users would notice a difference with consolidated metadata, I guess it would depend on their use-cases. Lazy metadata reads are great, but for sure it is faster to download a single small file than to download multiple files in a way that cannot be fully parallelized, even if they add up to the same total size.

**Comment by DennisHeimbigner** [comment]
> faster to download a single small file than to download multiple files 

true, but we have use cases where the client code is walking a large set of netcdf files and reading a few pieces of information out of each of them and where the total metadata is large (14 megabytes). This can occur when one has a large collection of netcdf files covering some time period and each netcdf file is a time slice (or slices).
Perhaps Rich Signell would like to comment with his experience.

**Comment by joshmoore** [comment]
> https://github.com/zarr-developers/zarr-specs/issues/41#issuecomment-833024978 I see this as very difficult. The reason is that the ncZarr conventions use files outside of the zarr hierarchy. We would probably need to implement a compatibility layer as a third-party package, similar to h5netcdf.

For what it's worth, I could see making some movement (June-ish?) on https://github.com/zarr-developers/zarr-specs/issues/112#issuecomment-825690209 to permit the additional files. But either way, certainly https://github.com/ome/ngff/pull/46#pullrequestreview-652899174 ([related issue](https://github.com/ome/ngff/issues/48/)) would suggest hammering out a plan for this difference before another package introduces a convention.

> https://github.com/zarr-developers/zarr-specs/issues/41#issuecomment-833036094 One thing I'll note on Xarray's convention for the Zarr is that we will likely change things in the near future to always write and expect "consolidated metadata" (see [pydata/xarray#5251](https://github.com/pydata/xarray/issues/5251)). This is almost completely backwards compatible, but if NcZarr writes these consolidated metadata fields in Xarray compat mode we could load these Zarr stores much quicker in Xarray.

Having gone through https://github.com/pydata/xarray/issues/5251 I'm slightly less worried about this then when I first read itI, but having just spent close to 2 months trying to get `dimension_separator` "standardized", I'd like to raise a flag that consolidated metadata is a similar gray area. It'd be nice to get it nailed down.

**Comment by rsignell-usgs** [comment]
@DennisHeimbigner, just a quick comment that I too always use consolidated metadata when writing Zarr.   Here's a recent example of some ocean model output we are publishing, where consolidated metadata is an order of magnitude faster to open: 
https://nbviewer.jupyter.org/gist/rsignell-usgs/bd10b623a90982303a4c0010e5db997d

**Comment by DennisHeimbigner** [comment]
Note that the issue for me is: for what use-cases is lazy metadata download better than consolidated metadata. The latter is better in the cases where you know that you need to access almost all of the meta-data or where the total size of the metadata is below some (currently unknown) size threshold. My speculation is that the access patterns vary all over the place and are highly domain dependent.  I infer that Rich's use case is one where all the metadata is going to be accessed.

In any case, once .zmetadata is well-defined (see Josh's previous comment) I will be adding it to nczarr. However, we will probably give the user the choice to use it or not if lazy download makes more sense for their use-case.

On the other side, it seems to me that zarr-python might profitably explore lazy download of the metadata.

**Comment by shoyer** [comment]
> Note that the issue for me is: for what use-cases is lazy metadata download better than consolidated metadata. The latter is better in the cases where you know that you need to access almost all of the meta-data or where the total size of the metadata is below some (currently unknown) size threshold. My speculation is that the access patterns vary all over the place and are highly domain dependent. I infer that Rich's use case is one where all the metadata is going to be accessed.

Agree! I'm sure there are cases where using consolidated metadata is not a great idea, though my guess is that they are relatively rare.

> In any case, once .zmetadata is well-defined (see Josh's previous comment) I will be adding it to nczarr. However, we will probably give the user the choice to use it or not if lazy download makes more sense for their use-case.

Sounds great, thanks!

> On the other side, it seems to me that zarr-python might profitably explore lazy download of the metadata.

As I understand it, this is already the case in Zarr-Python (if not using consolidated metadata). It's just that lazy metadata does not work for Xarray.

In particular, I think there is definitely a place for including an explicit "index" of arrays in a group that doesn't require potentially expensive directory listing. Hopefully this is already in the draft v3 spec (I haven't checked).

**Comment by DennisHeimbigner** [comment]
Can you elaborate on what you mean by this?
````
I think there is definitely a place for including an explicit "index" of arrays in a group that doesn't require potentially expensive directory listing.
````
Can you give an example?

**Comment by DennisHeimbigner** [comment]
Rich- are you using zarr-python directly or using xarray?

**Comment by shoyer** [comment]
> Can you elaborate on what you mean by this?
> 
> ```
> I think there is definitely a place for including an explicit "index" of arrays in a group that doesn't require potentially expensive directory listing.
> ```
> 
> Can you give an example?

I was thinking of exactly the information you store in `.nczgroup`.

**Comment by shoyer** [comment]
> Rich- are you using zarr-python directly or using xarray?

It looks like Rich is using Xarray here.

**Comment by DennisHeimbigner** [comment]
ok

**Comment by DennisHeimbigner** [comment]
I get the impression from various conversations that people do not like the nczarr convention of using extra objects like .nczarray.
As an alternative, we can store the same netcdf specific metadata as extra keys inside the various zarr standard objects. Thus most of the stuff in .nczarray could be inserted into .zarray with, say, a key named _nczarr_array.
My impression is that adding extra keys might be more acceptable than adding extra objects, but opinions welcome.

**Comment by joshmoore** [comment]
Thinking about
> I get the impression from various conversations that people do not like the nczarr convention of using extra objects like .nczarray.

and
> @shoyer https://github.com/zarr-developers/zarr-specs/issues/41#issuecomment-833184536 But if you do, Xarray will be able to read the data much faster.

I wonder if a .xarray file back in the day wouldn't have led to the need for consolidated metadata? i.e. are the two intertwined?

**Comment by DennisHeimbigner** [comment]
".xarray" -- I like it. Too bad it wasn't used. Still, it is not too late.

**Comment by shoyer** [comment]
> I get the impression from various conversations that people do not like the nczarr convention of using extra objects like .nczarray.

I think the separate objects is at least as inherently reasonable as custom attributes. There are clearly tradeoffs to both approaches.

Of course, it would be nice if we could use a shared standard, both:

1. for storing custom metadata needed by Zarr extensions, and
2. for representing dimension names on arrays

> I wonder if a .xarray file back in the day wouldn't have led to the need for consolidated metadata? i.e. are the two intertwined?

I wasn't involved in the discussion around adding consolidated metadata. I'm sure Xarray was one of the motivating use-cases, but performance concerns around opening lots of small files are definitely not unique to Xarray, as a search for [uses on GitHub](https://github.com/search?p=1&q=zarr.consolidate_metadata&type=Code) confirms. In my opinion, the concerns around performance with consolidated metadata and storing dimension names can be decoupled, so I would not favor combining them into `.xarray`.

If you look at the logic for `consolidate_metadata()`, it just gathers today metadata files found in a store:
https://github.com/zarr-developers/zarr-python/blob/ab5d91b62b83214d5a2b250d830d1996bb08cc56/zarr/convenience.py#L1112

producing a JSON document that looks something like:
```
{
    'zarr_consolidated_format': 1,
    'metadata': {
        '.zgroup': ...,
        '.zattrs': ...,
        'array/.zarray': ...,
        'subgroup/.zgroup': ...,
        ...
    }
}
```

It would be pretty reasonable to use the same approach for consolidating arbitrary metadata extensions stored in separate files, including all those used by NCZarr.

**Comment by joshmoore** [comment]
I'd definitely be for having a clear relationship between the consolidated metadata and the extra keys (https://github.com/zarr-developers/zarr-specs/issues/112#issuecomment-825690209) but I'll point out that they are guaranteed to be JSON, so we'd either need to _detect_ JSON and ingest or have a naming scheme for detectable "extension" files.

**Comment by rabernat** [comment]
Let me first state clearly that the technical discussion here and my comments below are not intended as a personal or institutional criticism of anyone; this discussion is a necessary part of the (sometimes slow / frustrating) community open source development process. I thank everyone for engaging with such patience and feel nothing but respect for everyone in the conversation.

The need for `.zmetadata` is most definitely not unique to Xarray. The ability of stores to "list" a directory was assumed in the V2 spec. However, certain types of stores (e.g. vanilla HTTP) are inherently unlistable. Other stores are slow to list, such as s3 buckets with hundreds of thousands of objects or directories on certain HPC filesystems.

`.zmetadata` mitigates these problems in two ways:
1. It explicitly enumerates all of the relevant Zarr metadata objects (`.zarray`, `.zattrs`, `.zgroup`) within a directory. This means that you never need to explicitly "list" the store.
2. It actually duplicates the contents of those files within a nested json structure. This reduces the number of distinct read operations you need to open a dataset (to one). This is important for high-latency stores, although the problem could also be mitigated with async.

The original consolidated metadata PR (https://github.com/zarr-developers/zarr-python/pull/268) contains more useful and relevant discussion.

> ".xarray" -- I like it. Too bad it wasn't used. Still, it is not too late.

I _don't_ like this idea and think it would have been a mistake. Xarray is not a data / metadata standard. It's an analysis library. Xarray's data model is [explicitly based on netCDF](http://xarray.pydata.org/en/stable/user-guide/data-structures.html#dataset): **"[xarray.Dataset] is designed as an in-memory representation of the data model from the netCDF file format."** Xarray made different choice from Unidata about how to encode netCDF into Zarr, but the end goal is the same.

> I get the impression from various conversations that people do not like the nczarr convention of using extra objects like .nczarray.

I am admittedly one of those people. However, I understand completely that when developing something new, certain technical decisions have to be made in order to bootstrap a project. I am familiar with this challenge, because I am part of the group (together with @jhamman and @shoyer) who made the original decision about how to encode / decode Xarray Datasets into Zarr.  Reviewing that original PR (https://github.com/pydata/xarray/pull/1528) is instructive. I have learned a lot since then, and probably would have made different choices.

At the time, we thought that the only thing we had to solve was: _how to we encode named dimensions in Zarr?_. We made the _ad hoc_ choice of using the [`_ARRAY_DIMENSIONS` key](https://github.com/pydata/xarray/blob/master/xarray/backends/zarr.py#L32
) within the standard array metadata (`.zattrs`). A different choice would have been to create an extra metadata object (a la nczarr); however, that would have been much harder to implement because zarr-python itself would not know about such objects, so we would have had to go around zarr-python. As it is, xarray's zarr implementation sits completely on top of zarr python and never interacts directly with the lower-level stores. In general, I still think this is a good model for higher-level libraries to follow.

Hindsight is 20/20. In retrospect, a better (but slower) path, would have been to work within the Zarr community to standardize an official convention for named dimensions. Maybe it would have been an extra metadata object. I personally think it would fit well in `.zarray` as an optional key. Maybe it would have been `_ARRAY_DIMENSIONS` in `.zattrs`. In any case, it would have been _an accepted standard_.

Now with the release of nczarr, we have **even more** _ad hoc_ conventions that were developed outside of the Zarr community process. For Unidata, there was never a choice to build _on top_ of an existing base Zarr implementation, because you were writing everything from scratch to begin with, so this distinction probably seemed academic; why not just add new features to Zarr while you're at it? However, I do wonder whether it would have been best to first build a Zarr C library with feature parity with zarr-python, rather than going directly to nczarr. (Similar to the relationship between HDF5 and netCDF.) 

In any case, in both cases the implementer (either Xarray devs or Unidata devs) went forward with a reasonable but unilateral technical decision about how to encode stuff on top of the base Zarr data model.  It's a cliche to post this cartoon, but...

![XKCD standards](https://imgs.xkcd.com/comics/standards.png)

The problem we all face now is: _how do implementations support an expanding set of ad hoc conventions?_  Is it too late too deprecate some of the conventions that are out there? There are [petabytes of data](https://aws.amazon.com/blogs/publicsector/now-available-cmip6-dataset-foster-climate-innovation-study-impact-future-climate-conditions/) in the wild using the xaray `_ARRAY_DIMENSIONS` convention, so I think that ship has sailed. And it's not very problematic, since the data are 100% compatible with vanilla Zarr. But what about nczarr? Are any data providers serving this format yet?

Going forward, I still think the best path is to _add optional named dimensions to the Zarr spec_, and then have downstream data models leverage this, rather than coming up with new conventions outside of the spec. Overall, it would be great to absorb the lessons of nczarr into the V3 spec process and provide a standard, community-consensus-based way to extend the spec in an open and transparent. way.

**Comment by rabernat** [comment]
A final comment on process.

The nczarr development has occurred openly on GitHub via PRs: https://github.com/Unidata/netcdf-c/pulls?q=is%3Apr+nczarr. There was an opportunity almost a year ago for us (Zarr + Xarray python community) to engage with the nczarr decisions (e.g. https://github.com/Unidata/netcdf-c/pull/1769) and provide feedback _before_ a release of nczarr. We missed this opportunity. It was a difficult year for everyone (to say the least! 😷), and a lot of balls got dropped. But going forward, let's try to find the bandwidth for this type of engagement and collaboration, particularly around the V3 spec.

**Comment by DennisHeimbigner** [comment]
First let me be clear that I have no problem with .zmetadata and Unidata will release  in NCZarr it as soon as it is standardized.
Second, it is still not too late to modify the NCZarr extensions since it is unlikely that it has been used for significant amounts of data.

**Comment by DennisHeimbigner** [comment]
One thing that surprises me is that the HDF5 community has not weighed in on this. Mapping HDF5 to Zarr is also going to need similar extensions as required by NetCDF-4.

**Comment by rabernat** [comment]
> One thing that surprises me is that the HDF5 community has not weighed in on this. Mapping HDF5 to Zarr is also going to need similar extensions as required by NetCDF-4.

I don't think that HDF5 / Zarr interoperability is a priority for HDFGroup. I think they have their hands full supporting their own format and products, and they have no incentive to support Zarr.

From the Zarr perspective, we have tried hard to make Zarr-python API-compatible with h5py, so users can easily swap one for another. And we provide [utilities to copy to / from HDF5](https://zarr.readthedocs.io/en/stable/tutorial.html#copying-migrating-data).

**Comment by DennisHeimbigner** [comment]
# New Representation of NCZarr extended metadata


After much discussion inside Unidata, and taking account of comments here, Unidata would like to propose a new representation for the NCZarr extensions to the Zarr format.

The new NCZarr extensions are stored as standard Zarr attributes in the .zattrs object. They use a naming convention to mark those attributes as NCZarr specific. The proposed naming convention is to prefix each attribute with the string ````_nczarr````. Unrecognized attributes prefixed with this string are ignored. But developers should be aware that Unidata will probably add new attributes with this prefix as support for NetCDF-4 is enhanced. 

The value of these attributes is intended to be an arbitrary JSON object. But Zarr only supports simple values or arrays of such. To bypass this constraint, an nczarr attribute's value consists of a single string value. This value is assumed to encode an arbitrary JSON object so the string value can be parsed to obtain the actual extended metadata. As a rule, the parsed value is a dictionary. Unrecognized keys are ignored to allow for future changes. 

## Proposed extension Attributes

The NCZarr extension metadata is stored as follows. 

### Extended .zgroup Attributes


### _nczarr_version
The root group stores the NCZarr version number. This is stored only in the root group. 

Example:
````
"_nczarr_version": "2.0.0"
````


### _nczarr_group


This contains a dictionary listing the names of variables (arrays), and subgroups contained (non-transitively) in a group. It also contains both the names and sizes of dimensions defined in that group. Note that eventually, the size field will allow some mechanism for specifying that the dimension has an "unlimited" size (in the NetCDF/HDF5 sense of unlimited). 

Example:
````
"_nczarr_group": "{
\"dimensions\": {\"d1\": \"1\", \"d2\": \"1\",...}
\"variables\": [\"v1\", \"v2\", ...]
\"groups\": [\"g1\", \"g2\", ...]
}"
````


### Extended .zarray Attributes


### _nczarr_array


This contains a dictionary listing the dimension references as fully qualified names. Ideally, we would use the XArray "_ARRAY_DIMENSIONS" convention, but it does not support fully qualified names. 

As an alternate representation, the "_ARRAY_DIMENSIONS" attribute could be extended to allow the storage of fully qualified names. As long as XArray is restricted to a single, root group, there would be no conflict. That is, a multigroup Zarr file is not XArray compatible, so adapting the contents of "_ARRAY_DIMENSIONS" presents no conflict. Using this alternative would, of course, require permission from the XArray group. It also depends on what other implementations do when they encounter "_ARRAY_DIMENSIONS" in a multi-group Zarr file. 

Example:
````
"_nczarr_array": "{
\"dimensions\": [\"/g1/g2/d1\", \"/d2\",...]
}"
````


### _nczarr_attrs


This contains a dictionary listing the attribute types. 

Example:
````
"_nczarr_attrs": "{
\"types\": {\"attr1\": \"<i4\", \"attr2\": \"<i1\",...}
}"
````

__Last Revised__: 5/15/2021

**Comment by shoyer** [comment]
> The value of these attributes is intended to be an arbitrary JSON object. But Zarr only supports simple values or arrays of such.

I don't think that's the case? I agree it's a little vague in the spec, but I'm pretty sure intent is to allow for storing arbitrary JSON documents in values, e.g., I'm pretty sure this in valid in the Zarr Python client. "Simple" is not a subset of JSON, it's a description of JSON itself 😀 

Assuming I'm right I would definitely suggest storing normal JSON documents. This is much more readable than JSON encoded strings.

**Comment by DennisHeimbigner** [comment]
Thanks for the info. If and when the spec clarifies, then it is worth revisiting.

**Comment by shoyer** [comment]
> This contains a dictionary listing the dimension references as fully qualified names. Ideally, we would use the XArray "_ARRAY_DIMENSIONS" convention, but it does not support fully qualified names.
> 
> As an alternate representation, the "_ARRAY_DIMENSIONS" attribute could be extended to allow the storage of fully qualified names. As long as XArray is restricted to a single, root group, there would be no conflict. That is, a multigroup Zarr file is not XArray compatible, so adapting the contents of "_ARRAY_DIMENSIONS" presents no conflict. Using this alternative would, of course, require permission from the XArray group. It also depends on what other implementations do when they encounter "_ARRAY_DIMENSIONS" in a multi-group Zarr file.

I'm not entirely opposed to fully qualified names, but I'm not sure I agree we need them.

First, to clarify: Xarray can actually load datasets from Zarr groups, but only from a single group at a time.

As I understand the netCDF4 data model, dimensions are visible to the arrays in a group and all sub-groups. Duplicate dimensions at the same level are not allowed. It is allowed to override dimensions in a sub-group. So far, this is all compatible with unqualified dimension names, with the "nearest" name taking precedence.

Are you concerned about cases where a dimension has a different size in a sub-group, but a variable in that sub-group uses a dimension of the same name from a parent group? If this is indeed possible in netCDF-C, I sincerely hope that there are no files that use this feature out in the wild. In my mind, this violates an implicit assumption of the netCDF data model, which is that uses of a dimension name among variables in a group have the same size. These files would be incompatible with Xarray and I suspect many other tools for working with netCDF data. As far as I can tell, it is not possible to create such netCDF files using netCDF4-Python.

**Comment by DennisHeimbigner** [comment]
This is incorrect. The search upwards rule only applies to the CDL representation.
It is perfectly legal to use any dimension for a variable, including one  arbitrarily
deep in some other group. So this is why FQNs are required.

**Comment by shoyer** [comment]
> This is incorrect. The search upwards rule only applies to the CDL representation.
> It is perfectly legal to use any dimension for a variable, including one arbitrarily
> deep in some other group. So this is why FQNs are required.

From the netCDF documentation on groups: https://www.unidata.ucar.edu/software/netcdf/docs/groups.html
"Dimensions are scoped such that they are visible to all child groups. For example, you can define a dimension in the root group, and use its dimension id when defining a variable in a sub-group."

In my opinion, this is a good constraint to enforce, for the sake of user sanity :). Even if netCDF4 doesn't enforce it, we could enforce it for NCZarr.

**Comment by czender** [comment]
I was just pointed to this discussion by a helpful colleague. Regarding dimensions that are _not_ defined in groups that are direct ancestors of the group(s) where they are used: As Dennis says, netCDF4 allows it. Moreover, NASA stores a large amount of data in the HDF5-EOS format. Some of those datasets (e.g., the OMI instrument on the Aura satellite) put the geophysical fields (e.g., ozone) in groups (e.g., sub-groups) that are cousins (not children) of the groups where the spatial/temporal dimensions are defined. I hope a solution is found that is both user/developer-friendly and ensures that NCZarr has the flexibility it needs to represent and access such datasets.
I do agree it is a poor practice to for data producers to organize data in ways that make direct ancestor searches insufficient to resolve all the metadata values. 
FYI: The CF Metadata Convention formally adopted Groups as part of CF-1.8 in 2019. However, CF-compliant datasets may only use dimensions defined in direct ancestors of the groups where they are referenced. In other words, CF does not support the full generality of dimension references supported by netCDF4, such as the HDF5-EOS datasets mentioned above.

**Comment by shoyer** [comment]
@czender thanks for joining us! Interesting to hear that non-hierarchical dimensions are in fact widely used. So it sounds like NCZarr needs this feature.

It sounds like this needs a new extension point like the proposed ​`_nczarr_array` attribute. The problem is that there are likely already Xarray datasets that have been written to Zarr groups without fully qualified names, and there is also other application code (e.g., [neuroglancer](https://github.com/google/neuroglancer)) that expects the entries in `_ARRAY_DIMENSIONS`  to be plain names.

That said, I do still think there is value in having a "domain agnostic" way to represent dimension names on arrays, without requiring associated dimension objects. Non-geoscience libraries like Neuroglancer may not care enough to implement the full NCZarr standard, but they would be willing to handle basic dimension names. 

Ideally NCZarr could be a super-set of this domain agnostic functionality, e.g., perhaps by writing  the `_ARRAY_DIMENSIONS` attribute even if it is redundant with `_nczarr_array`.

**Comment by shoyer** [comment]
Another thought: perhaps it would make sense to name these NCZarr attributes in all capitals like `_NCZARR_ARRAY` (rather than lowercase). This would be a little more consistent with Xarray's `_ARRAY_DIMENSIONS` naming convention, and would distinguish them more clearly from typical user provided attributes.

**Comment by DennisHeimbigner** [comment]
I have been working on how to allow nczarr to interoperate with zarr numcodec JSON filter definitions. I have a blog post of a proposal. It is  a bit netcdf-c centric, but some in the zarr community might find it interesting.
````
https://www.unidata.ucar.edu/blogs/developer/en/entry/nczarr-support-for-zarr-filters
````

**Comment by rouault** [comment]
I've given a try at netcdf-c master with the Version 2 NCZarr Extented Meta-data, and I think it is slightly non conformant with the Zarr V2 specification

I've converted a netcdf file to the new NCZarr format with nccopy and I see in a .zarray file
```json

{
  "zarr_format": 2,
  "shape": [
    20,
    20
  ],
  "dtype": "<i1",
  "chunks": [
    20,
    20
  ],
  "order": "C",
  "compressor": null,
  "filters": null,
  "_NCZARR_ARRAY": {
    "dimrefs": [
      "/y",
      "/x"
    ],
    "storage": "chunked"
  }
}
```

The presence of the ``_NCZARR_ARRAY`` key seems to contradict this point of the spec in https://zarr.readthedocs.io/en/stable/spec/v2.html#metadata : "Other keys SHOULD NOT be present within the metadata object and SHOULD be ignored by implementations."

**Comment by DennisHeimbigner** [comment]
There was a long conversation about this on the Zarr spec github.
I pointed out that the new "dimension_separator" key violated
this constraint. The consensus seemed to be that extra keys would be allowed,
but must be ignored if they are not recognized by the implementation.
I have not checked to see if that change has made it into the spec yet.

**Comment by shoyer** [comment]
I agree, I think any reasonable implementation should *ignore* unrecognized
keys or files. Hopefully this will be codified in zarr v3.

**Comment by rouault** [comment]
> The consensus seemed to be that extra keys would be allowed, but must be ignored if they are not recognized by the implementation.

ok, thanks for the clarification

**Comment by joshmoore** [comment]
for reference: https://github.com/zarr-developers/zarr-python/pull/715#issuecomment-821094669

That didn't make it into a zarr-specs issue (neither v2 nor v3) as far as I can tell. Anyone up for shepherding that?

**Comment by joshmoore** [comment]
See the related conversation in https://github.com/pydata/xarray/issues/6374

**Comment by halehawk** [comment]
@DennisHeimbigner does NCZarr support any filter now?

**Comment by DennisHeimbigner** [comment]
Yes although there are some complications because the code uses HDF5 filters to perform
the actual filter code, but it needs extra code to convert a Zarr codec JSON format
to the HDF5 unsigned integer parameters.
What specific filter(s) do you need?

**Comment by halehawk** [comment]
@DennisHeimbigner Do you have documentations about how to enable and use the filter through NCZarr, and we have a new codec which is not binding to any codec yet. Do you have suggestion on how to enable it in your NCZarr?

**Comment by shaomeng** [comment]
@DennisHeimbigner @halehawk  Maybe I should jump in now ;)

I have a lossy compressor product (SPERR: https://github.com/shaomeng/SPERR) that I'm looking at paths to integrate into the Zarr format. I haven't spent too much time on it, but my understanding is that I need to make it a Zarr filter. Our immediate application of it, an ASD run of MURam, has decided to use NCZarr to  output Zarr files, so the question arises that if Zarr filters are supported by NCZarr.

I guess the most direct question to @DennisHeimbigner as the NCZarr developer is, what approach do you recommend to integrate a lossy compressor to an NCZarr output?

**Comment by DennisHeimbigner** [comment]
If the compressor is (or easily could be) written in python, then see the NumCodecs web page.
If the compressor is in C or C++, and you decide to use netcdf-c NCZarr, then you need
to build an HDF5 compressor wrapper plus the corresponding codecs API.
I have attached the relevant documentation. If this compressor is similar to some existing
compressor such as bzip2 or zstandard, then you can copy and modify the corresponding
wrapper in netcdf-c/plugins directory -- H5Zzstd.c, for example.
[filters.md](https://github.com/zarr-developers/zarr-specs/files/8455495/filters.md)

**Comment by shaomeng** [comment]
That's super helpful, thanks for the pointer! One more question: can the
compression-enabled NCZarr output read by zarr tools in the python
ecosystem?

**Comment by DennisHeimbigner** [comment]
Yes, IF the filters are available in NumCodecs.

**Comment by halehawk** [comment]
Does it mean the compressor would be better to be integrated in numcodecs if it wants to be used by nczarr and zarr/xarray?

Haiying

**Comment by DennisHeimbigner** [comment]
Sorry, I wasn't clear.
Suppose you use nczarr to write a Zarr file where some of its arrays apply a filter.
Then you can obviously read that file with nczarr.
However, suppose you write the array with nczarr and then want
others to read it using python-zarr. In that case, you will need to create 
a NumCodecs compatible version of your filter written in python so that
the python-zarr users can read the data written by nczarr.

**Comment by DennisHeimbigner** [comment]
There are two pieces here and I am sorry I was unclear.
The first piece is the declaration of the compressor for a variable
in the Zarr metadata, This is specified in the "compressor" key for
the .zarray metadata object for the variable. The format for this
is defined by NumCodecs and generally has the form
````
{"id": "<compressor name>", "parameter1": <value>, ... "parametern": <value>}
````
So for zstd, we might have this:  ````{"id": "zstd", "level:": 5}````

The second part is the actual code that implements the compressort.

NCZarr supports the first part so that it can read/write legal Zarr metadata.
BUT, NCZarr requires its filter code to be written in C (or C++).
More specifically, it does not support the Python compressor code implementations.
Sorry for the confusion

**Comment by halehawk** [comment]
@Dennis Heimbigner ***@***.***> I still need you to clarify something
here. So I looked at your H5Zzstd.c which is a HDF5 plugin for zstd and
supports numcodec zstd read/write. Then I got this idea, if Samuel's new
compressor need not get a formal HDF5 filter ID, but should add a similar
H5Zsperr.c to your netcdf-c if he need not write to HDF5/netcdf file now.
But he needs to be in numcodecs to get Zarr/numcodecs support.

**Comment by shaomeng** [comment]
> NCZarr requires its filter code to be written in C (or C++). 

Just to clarify, did you mean that NCZarr requires its filter code to be in C *AND* also exposed to NCZarr as HDF5 filters? E.g., NumCodecs filters won't work. 

Sorry for the back and forth in this github thread. I think this is my last try and if there's still confusion, I'll try to set up a meeting and resolve it more directly :)

**Comment by jbms** [comment]
I don't know the details of how codecs are defined for NCZarr, but in general you will need to provide a separate implementation of a codec for each zarr implementation that you want to support it.

Zarr-python provides a mechanism by which codecs can be registered --- numcodecs defines many codecs, and zarr-python pulls in numcodecs as a dependency, but it is actually possible to define a codec for zarr-python outside of the numcodecs package --- see for example the imagecodecs Python package.

**Comment by DennisHeimbigner** [comment]
> @dennis Heimbigner ***@***.***> I still need you to clarify something here. So I looked at your H5Zzstd.c which is a HDF5 plugin for zstd and supports numcodec zstd read/write. Then I got this idea, if Samuel's new compressor need not get a formal HDF5 filter ID, but should add a similar H5Zsperr.c

That is correct. THe HDF group reserves ids 32768 to 65535 for unregistered use.
So Samuel can pick a number in that range for his filter; later, if desired, a forml
HDF5 filter id can be assigned.

**Comment by joshmoore** [comment]
First a big :100: for the discussion, since this is exactly what we want to see happening for cross-implementation support of codecs. @shaomeng & @halehawk, don't hesitate to keep asking. 

I do wonder, @DennisHeimbigner, if we don't want to establish the channel you'd like for more nczarr questions. If so, I'd say we update the start and end of this thread with that URL and close this issue.

Others may want to express an opinion, but if it's useful, we can have a no-code location like github.com/zarr-developers/nczarr for people to find a README pointing to the netcdf-c implementation's resources.

cc: @WardF

**Comment by briannapagan** [comment]
[21-050r1_Zarr_Community_Standard.pdf](https://github.com/zarr-developers/zarr-specs/files/11334811/21-050r1_Zarr_Community_Standard.pdf)
Adding this here for reference in the convo.

**Comment by dblodgett-usgs** [comment]
Pertinent text from @briannapagan's link above...

> Beginning with NetCDF-C version 4.8.0, Unidata introduced experimental Zarr support
into the NetCDF-C library. This was accomplished via creating a new specification -
NCZarr - which is “similar to, but not identical with the Zarr Version 2 Specification.”
Specifically, NCZarr adds two additional metadata files (“.nczarray" and ".nczattr”),
which are not part of the Zarr V2 Spec. Since NCZarr stores are not fully compatible and
interoperable with Zarr V2, this community standard excludes NCZarr. Work is ongoing
to reconcile NCZarr and the architectural reasons that motivated its development with the
forthcoming Zarr V3 Specification.
Fortunately, the NetCDF-C library also supports reading / writing of data using the
simpler Named Dimension convention described in 4.1.

**Comment by DennisHeimbigner** [comment]
That information is out-of-date in a couple of ways.
1. the metadata files  (“.nczarray" and ".nczattr”)  are no longer used; they were replaced with special dictionary entries.
2. I believe the spec was changed to specify that unrecognized elements (objects and dictionary entries) should be ignored by any implementation that does not recognize them.
3. With \#2 in effect, nczarr created files can be read by pure zarr implementations and nczarr can read pure zarr files.

**Comment by dblodgett-usgs** [comment]
Thanks for calling that out, @DennisHeimbigner. This came out of a conversation over here. https://github.com/zarr-developers/geozarr-spec/issues/22 

There are very few people who have a deep enough understanding of the moving parts here to answer all the questions. It's good to hear that we basically have interoperability. 

Two questions:

1. Do you feel like we even need to worry about the distinction right now? 
2. Is there a current document we should be using to learn about the nuances between "pure zarr" and "nczarr"?
